test/clock.test.js

```diff
@@ -212,4 +220,67 @@ describe('clock', () => {
     assert.equal(head.length, 1)
     assert.equal(head[0].toString(), event1.cid.toString())
   })
+
+  it('test that contains only traverses history once', async () => {
+    const blocks = new Blockstore()
+    const root = await EventBlock.create(await randomEventData())
+    await blocks.put(root.cid, root.bytes)
+    /** @type {API.EventLink<any>[]} */
+    let head = [root.cid]
+    const parents0 = head
+    const blockGet = blocks.get.bind(blocks)
+    let count = 0
+    blocks.get = async cid => {
+      count++
+      return blockGet(cid)
+    }
+    const event0 = await EventBlock.create(await randomEventData(), parents0)
+    await blocks.put(event0.cid, event0.bytes)
+    head = await advance(blocks, head, event0.cid)
+
+    const event0a = await EventBlock.create(await randomEventData(), [event0.cid])
+    await blocks.put(event0a.cid, event0a.bytes)
+    head = await advance(blocks, head, event0a.cid)
+
+    const event0b = await EventBlock.create(await randomEventData(), [event0a.cid])
+    await blocks.put(event0b.cid, event0b.bytes)
+    head = await advance(blocks, head, event0b.cid)
+    const event0c = await EventBlock.create(await randomEventData(), [event0b.cid])
+    await blocks.put(event0c.cid, event0c.bytes)
+    head = await advance(blocks, head, event0c.cid)
+
+    const event1 = await EventBlock.create(await randomEventData(), [event0c.cid])
+    await blocks.put(event1.cid, event1.bytes)
+    head = await advance(blocks, head, event1.cid)
+
+    const event2 = await EventBlock.create(await randomEventData(), parents0)
+    await blocks.put(event2.cid, event2.bytes)
+    head = await advance(blocks, head, event2.cid)
+
+    const event3 = await EventBlock.create(await randomEventData(), [
+      event0c.cid,
+      event1.cid,
+      event2.cid
+    ])
+    await blocks.put(event3.cid, event3.bytes)
+    head = await advance(blocks, head, event3.cid)
+
+    const event4 = await EventBlock.create(await randomEventData(), [event0c.cid, event1.cid])
+    await blocks.put(event4.cid, event4.bytes)
+    head = await advance(blocks, head, event4.cid)
+
+    const event5 = await EventBlock.create(await randomEventData(), [event3.cid, event4.cid])
+    await blocks.put(event5.cid, event5.bytes)
+    head = await advance(blocks, head, event5.cid)
+
+    const event6 = await EventBlock.create(await randomEventData(), parents0)
+    await blocks.put(event6.cid, event6.bytes)
+    const before = count
+    head = await advance(blocks, head, event6.cid)
```

Comment:
I think we could set this test up a little simplier and more realistically. The event chain you've created looks like this:

```
                                                   ┌───────────────────┬──────┐
                                                   │                   │event4│◄──────┐
                                                   │             ┌─────┴──────┘       │
                                                   │             │                    │
                                                   │        ┌────▼─┐                  │
                                                   │  ┌─────┤event1◄──────┐       ┌───┴──┐
                                                   │  │     └──────┘      │       │event5│
           ┌──────┐     ┌───────┐   ┌───────┐   ┌──▼──▼─┐                 │       └───┬──┘
    ┌──────┤event0│◄────┤event0a│◄──┤event0b│◄──┤event0c│                 │           │
    │      └──────┘     └───────┘   └───────┘   └─────▲─┘                 │           │
    │                                                 │                ┌──┴───┐       │
    │                                                 └────────────────┤event3│◄──────┘
    │                                                                  └──┬───┘
┌───▼───┐  ┌──────┐                                                       │
│genesis│◄─┤event2◄───────────────────────────────────────────────────────┘
└───▲───┘  └──────┘
    │
    │
    │      ┌──────┐
    └──────┤event6│
           └──────┘
```

`event3` and `event4` would _never_ have parents `event0c` _and_ `event1` in a properly functioning clock. I think it would be less confusing to fix this issue at least.

I assume (hope) this isn't based on an actual clock in the wild!

We arrive at 13, because when we advance with `head: [event5], event: event6` we firstly check if the head is in the event:

* get `event5` (1)
* get `event6` (2)
* get `genesis` (3)

Then we check if the event is in the head, with de-duping:

* get `event6` (4)
* get `event5` (5)
* then get each of the nodes back to parent(s) of `event6` (`genesis`)
    * `event4` (6)
    * `event3` (7)
    * `event0c` (8)
    * `event1` (9)
    * `event0b` (10)
    * `event0a` (11)
    * `event0` (12)
    * `genesis` (13)

Reply:
p.s. https://asciiflow.com is great

Reply:
That is b/c it they know about event 1 which has another path to event0c. the naming is bad, but we can try to substitute `event6` as the other ancestor instead of `event0c` and then fix the names

Reply:
here is the original as a mermaid flow:

```mermaid
flowchart TB
    genesis("genesis") --> event2("event2")
    genesis --> event6("event6")
    event2 --> event0("event0")
    event0 --> event0a("event0a")
    event0a --> event0b("event0b")
    event0b --> event0c("event0c")
    event0c --> event1("event1")
    event0c --> event3("event3")
    event1 --> event3
    event1 --> event4("event4")
    event4 --> event5("event5")
    event5 --> event3
```

and here is the new one, we've updated the code to move event6 earlier. 

```mermaid
flowchart TB
    root("root") --> event0("event0")
    event0 --> event0a("event0a")
    event0a --> event0b("event0b")
    event0b --> event0c("event0c")
    event0c --> event1("event1")
    event1 --> event2("event2")
    event1 --> event6("event6")
    event2 --> event3("event3")
    event6 --> event4("event4")
    event4 --> event5("event5")
```

Maybe worth trying something like:

```mermaid
flowchart TB
    root("root") --> event0("event0")
    event0 --> event0a("event0a")
    event0a --> event0b("event0b")
    event0b --> event0c("event0c")
    event0c --> event1("event1")
    event0 --> event1a("event1a")
    event1a --> event1
    event1 --> event2("event2")
    event1 --> event6("event6")
    event2 --> event3("event3")
    event6 --> event4("event4")
    event4 --> event5("event5")
```

Reply:
in the wild it's databases with multiple heads that cause the issue, maybe we need n > 2 to see it?

Reply:
Oh nice mermaid is better!